### ISSUE TITLE
jsx-one-expression-per-line should ignore inline elements

### ISSUE THREAD

**Issue by anthonator** [yannickcr/eslint-plugin-react#1848]
It doesn't seem to make sense to one-line inline elements.

For example, one-lining the `em` element in the example below doesn't make sense.

```js
<Styled.PasswordInstructions errored={ !!formErrors.userPassword }>
  Passwords must be at least 8 characters long and can&apos;t be
  things like <em>password</em>, <em>123456</em> or <em>abcdef</em>.
</Styled.PasswordInstructions>
```

Here is the "fixed" version.

```js
<Styled.PasswordInstructions errored={ !!formErrors.userPassword }>
  Passwords must be at least 8 characters long and can&apos;t be
  things like
  <em>
    password
  </em>
  ,
  <em>
    123456
  </em>
  or
  <em>
    abcdef
  </em>
  .
</Styled.PasswordInstructions>
```

I could also see an option allowing the developer to choose which elements to ignore along with an `ignoreInline` option.

**Comment by ljharb** [comment]
In that particular case i agree it looks messy; I’m not sure how to generically allow it when it’s inline with text without having things like long links be inlined.

**Comment by ljharb** [comment]
See #1855; I think they're related.

**Comment by ryansully** [comment]
Somewhat, but sort of a different concern. I left feedback there as well.

**Comment by Mindaugas-Jacionis** [comment]
@ljharb the reference to #1855 is confusing, because there you point back to this thread 😄

**Comment by ljharb** [comment]
@Mindaugas-Jacionis two things that relate to each other, cross-referencing each other, shouldn't be confusing.

**Comment by MrHen** [comment]
I recently had to disable `react/jsx-one-expression-per-line` because it made things like this much harder to read:

```
        <div>
            <span>ABC</span>
            { one && <span>DEF</span> }
            <span>GHI {two} JKL</span>
        </div>
```

Result:

```
        <div>
            <span>
                ABC
            </span>
            { one &&
                <span>
                    DEF
                </span>
            }
            <span>
                GHI
                {' '}
                {two}
                {' '}
                JKL
            </span>
        </div>
```

What I _want_ to prevent is someone trying to inline nested tags. I don't care about people using `<span>ABC</span>` -- that's easy enough to read. I can understand why the lint rule current does what it does but it would be handy to have an option to ignore inlined non-tag expressions.

**Comment by danny-andrews-snap** [comment]
Agreed. In addition, running `--fix` with this rule enabled completely mangles your code.

**Comment by igorpupkinable** [comment]
This had messed up our pages with content. Text is missing meaningful spaces.

**Comment by evenfrost** [comment]
Any update on this?

**Comment by ljharb** [comment]
See https://github.com/yannickcr/eslint-plugin-react/issues/1848#issuecomment-400011655

**Comment by Davidhidalgo** [comment]
This is very annoying... :disappointed:

**Comment by Guiga303** [comment]
> I recently had to disable `react/jsx-one-expression-per-line` because it made things like this much harder to read:
> 
> ```
>         <div>
>             <span>ABC</span>
>             { one && <span>DEF</span> }
>             <span>GHI {two} JKL</span>
>         </div>
> ```
> 
> Result:
> 
> ```
>         <div>
>             <span>
>                 ABC
>             </span>
>             { one &&
>                 <span>
>                     DEF
>                 </span>
>             }
>             <span>
>                 GHI
>                 {' '}
>                 {two}
>                 {' '}
>                 JKL
>             </span>
>         </div>
> ```
> 
> What I _want_ to prevent is someone trying to inline nested tags. I don't care about people using `<span>ABC</span>` -- that's easy enough to read. I can understand why the lint rule current does what it does but it would be handy to have an option to ignore inlined non-tag expressions.

how did you that?

**Comment by nomasprime** [comment]
Agree this can hinder readability and an allow inline option is needed. Disabling rule for now.

**Comment by nidalR** [comment]
At least tags like <b> and <em> should be ignored to start with...

**Comment by TSMMark** [comment]
What's the solution? Add an option like `ignoreElements: ['em', 'p']` and/or `ignoreProps: ['key', 'whatever']` ?

**Comment by ljharb** [comment]
I'm not sure we have a solution yet - but it'd be great if someone wanted to propose one.

**Comment by TSMMark** [comment]
Great point, it's true that one of the best things about open source packages that other people wrote for you for free is that they are purely optional and you are not actually forced to use them! A lot of people don't know that. Thanks for notifying everyone subscribed to this thread with that valuable information!

Although I wouldn't recommend your approach for globally disabling an eslint rule across your entire codebase — you can simply add one line to your .eslintrc file inside the `rules` object, like:

```
"react/jsx-one-expression-per-line": 0
```

**Comment by ljharb** [comment]
I’m going to hide all three of these comments; disabling an eslint rule is well-understood and we’ll-documented, and is in no way a “solution”.

**Comment by jarommadsen** [comment]
Chiming in 7 years later to voice that this is still annoying lol.

I'd be in favor for an option to disable it for basic html element tags like `em`, `b`, `i`, etc.
